build.gradle

```diff
@@ -646,25 +646,29 @@ def calculateVersion() {
 }
 
 def getCheckedOutGitCommitHash() {
-  def gitFolder = "$projectDir/.git/"
-  if (!file(gitFolder).isDirectory()) {
-    // We are in a submodule.  The file's contents are `gitdir: <gitFolder>\n`.
-    // Read the file, cut off the front, and trim the whitespace.
-    gitFolder = file(gitFolder).text.substring(8).trim() + "/"
+  try {
+    def gitFolder = "$projectDir/.git/"
+    if (!file(gitFolder).isDirectory()) {
+      // We are in a submodule.  The file's contents are `gitdir: <gitFolder>\n`.
+      // Read the file, cut off the front, and trim the whitespace.
+      gitFolder = file(gitFolder).text.substring(8).trim() + "/"
+    }
+    def takeFromHash = 8
+    /*
+     * '.git/HEAD' contains either
+     *      in case of detached head: the currently checked out commit hash
+     *      otherwise: a reference to a file containing the current commit hash
+     */
+    def head = new File(gitFolder + "HEAD").text.split(":") // .git/HEAD
+    def isCommit = head.length == 1 // e5a7c79edabbf7dd39888442df081b1c9d8e88fd
+
+    if (isCommit) return head[0].trim().take(takeFromHash) // e5a7c79edabb
+
+    def refHead = new File(gitFolder + head[1].trim()) // .git/refs/heads/master
+    refHead.text.trim().take takeFromHash
+  } catch (Exception e) {
+    return "xxxxxxxx"
```

Comment:
It would be worth printing a message here so that if an unexpected exception is thrown in the normal code there’s some info about why.

Reply:
The stack trace is 257 lines long, so I hid it at "info" level and memoized the execution so you get it only twice not 20+ times.

Reply:
+1 seems like a good trade off.